src/decode-store.c

```diff
@@ -0,0 +1,320 @@
+/* Copyright (C) 2007-2019 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#include "suricata-common.h"
+#include "suricata.h"
+#include "conf.h"
+#include "decode.h"
+#include "decode-store.h"
+
+static inline struct DecodeStore *GetActiveStore(struct DecodeStore *s)
+{
+    while (s && s->is_full) {
+        s = s->next;
+    }
+    return s;
+}
+
+static inline struct DecodeStore *GetNewStore(void)
+{
+    struct DecodeStore *s = SCCalloc(1, sizeof(*s));
+    if (s == NULL)
+        return NULL;
+    return s;
+}
+
+static inline void Copy(struct DecodeStore *s, const uint8_t *data, size_t data_size)
+{
+    memcpy(s->data+s->data_offset, data, data_size);
+    s->data_offset += data_size;
+    s->is_full = (s->data_offset == DECODE_STORE_SIZE - 1);
+}
+
+static void PrintInfo(struct DecodeStore *s)
+{
+    SCLogDebug("store %p used %u out of %u", s, s->data_offset, DECODE_STORE_SIZE);
+}
+
+static void Append(Packet *p, uint8_t *data, size_t data_size)
+{
+    SCLogDebug("start");
+    struct DecodeStore *s = GetActiveStore(&p->decode_store);
+    SCLogDebug("s %p", s);
+    if (!s) {
+        s = GetNewStore();
+        if (!s) {
+            // SET EVENT?
+            return;
+        }
+    }
+    SCLogDebug("s %p", s);
+    if (data_size + s->data_offset < DECODE_STORE_SIZE) {
+        Copy(s, data, data_size);
+        SCLogDebug("data fits, done");
+        PrintInfo(s);
+        return;
+    }
+    SCLogDebug("data does not fit, get additional store");
+
+    s->data[s->data_offset] = DECODE_STORE_TYPE_NEXT;
+    s->is_full = true;
+
+    s->next = GetNewStore();
+    if (!s->next) {
+        // SET EVENT?
+        return;
+    }
+    s = s->next;
+    Copy(s, data, data_size);
+    SCLogDebug("got additional store. Done");
+    PrintInfo(s);
+}
+
+
+typedef union DecodeStoreBuiltinVLAN {
```

Comment:
Not liking the "Builtin" in the name too much.  "DecoderStoreTypeVlan", "DecodeStoreTypeGeneric", etc.